### PR TITLE
🌱 Automating release creation

### PR DESCRIPTION
**Pull request by adilGhaffarDev**
**What this PR does / why we need it**:
This PR is
- Fixing release automation according to this issue: https://github.com/metal3-io/cluster-api-provider-metal3/issues/1923 
- fixing release note generation
- updating documentation
**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

### CODE REVIEW

**Comment by adilGhaffarDev** [comment]
documentation needs to be updated I am working on it.

File: Makefile
Comment (tuminoid) [inline Makefile:309]: typo

File: .github/workflows/release.yaml
Comment (tuminoid) [inline .github/workflows/release.yaml:36]:
```suggestion
              echo "RELEASE_VERSION=${RELEASE_VERSION}" >> ${GITHUB_ENV}
```
Fix all the curlys in this file please.

File: .github/workflows/release.yaml
Comment (tuminoid) [inline .github/workflows/release.yaml:106]: This implies all changelogs are put into main, and not respective release branches. I guess that works as well.

File: Makefile
Comment (tuminoid) [inline Makefile:294]: Does this dir actually matter what its called?

File: .github/workflows/release.yaml
Comment (tuminoid) [inline .github/workflows/release.yaml:12]: Do add it to top-level permissions. Put it into specific job that needs that permission.

File: .github/workflows/release.yaml
Comment (adilGhaffarDev) [reply .github/workflows/release.yaml]: fixed

File: .github/workflows/release.yaml
Comment (adilGhaffarDev) [reply .github/workflows/release.yaml:106]: Yes PR will be opened and merge on the main branch.

File: Makefile
Comment (adilGhaffarDev) [reply Makefile:293]: No, it does not matter, we have to just make sure that this folder name is the same as the one we are checking in the workflow script. If we change it here we have to change in workflow script too. 
In my opinion CHANGELOG is ok, but we can change it if you have some other suggestions.

File: .github/workflows/release.yaml
Comment (adilGhaffarDev) [reply .github/workflows/release.yaml:12]: there are 2 jobs
- push_release_tags
- release
And both need this permission. 
Can steps have different permissions?

Comment (tuminoid) [reply .github/workflows/release.yaml:12]: Yes. Put minimum per job.

Adding to top-level permissions leads to 
1. excessive permissions when job needs differ
2. applies those top-level permissions to new jobs added which might not need any

Permissions is thus best applied per job.

File: Makefile
Comment (tuminoid) [reply Makefile:293]: Well, we have releasenotes dir already, which currently is just a placeholder. IMO it is just as valid name as changelog would be, without requiring any change here.

File: .github/workflows/release.yaml
Comment (tuminoid) [inline .github/workflows/release.yaml:68]: Let's make it clear what it tries to achieve.

```suggestion
          if [[ -n "$(git tag -l "${RELEASE_VERSION}")" ]]; then
```

File: .github/workflows/release.yaml
Comment (adilGhaffarDev) [reply .github/workflows/release.yaml]: done

File: Makefile
Comment (adilGhaffarDev) [reply Makefile]: changed it back to `releasenotes`, also removed `releasenotes` from gitignore.

File: .github/workflows/release.yaml
Comment (tuminoid) [inline .github/workflows/release.yaml:109]: this is still pointing to CHANGELOG

File: hack/tools/go.mod
Comment (tuminoid) [inline hack/tools/go.mod:5]: We don't want toolchain directives. Just put 1.22.9 directly to go directive, and remove this line.

Comment (tuminoid) [reply hack/tools/go.mod:5]: We should also bump go to 1.22.9 or 1.23.x. I think generate is failing because of that.

Comment (adilGhaffarDev) [reply hack/tools/go.mod]: let's do this in separate PR, I was able to fix the issue without needing to do that.

File: .github/workflows/release.yaml
Comment (adilGhaffarDev) [reply .github/workflows/release.yaml]: fixed

**Comment by tuminoid** [comment]
Also, let's design this so that we can replicate it to all releasable repos here. For other workflows, we have moved the boilerplate code to project-infra and then simply call it from repo being released with correct params. Makes much less maintenance in the long run, when updates need to go in one repo only. 

Here we of course have the release note code in the repo, which might make it difficult, but at least check if above is possible?

**Comment by adilGhaffarDev** [comment]
> Also, let's design this so that we can replicate it to all releasable repos here. For other workflows, we have moved the boilerplate code to project-infra and then simply call it from repo being released with correct params. Makes much less maintenance in the long run, when updates need to go in one repo only.

Yes, @mquhuy was also suggesting that. I will check.

**Comment by adilGhaffarDev** [comment]
As discussed, let's proceed with merging this change and test it in the upcoming patch releases. Once verified, I'll open a pull request in project-infra to add a common workflow script for IPAM, CAPM3, and BMO.
cc @tuminoid @Rozzii

**Comment by adilGhaffarDev** [comment]
@peppi-lotta @smoshiur1237 @Sunnatillo please review

File: .github/workflows/release.yaml
Comment (tuminoid) [inline .github/workflows/release.yaml:7]: Indents in the file are wrong, please fix.

```suggestion
    - main
```

File: .github/workflows/release.yaml
Comment (tuminoid) [inline .github/workflows/release.yaml:48]: This makes us commit to releasing where alpha and beta are tags on main, but RC and actual release tag has to be in release branch.

Do we REALLY want to commit to?

File: .github/workflows/release.yaml
Comment (tuminoid) [inline .github/workflows/release.yaml:77]: These tags are wrong?

They should match per repo, and for IPAM we have:
- main tag
- api tag
- test tag

File: .github/workflows/release.yaml
Comment (adilGhaffarDev) [reply .github/workflows/release.yaml:48]: I think this is fine, this is something that is followed in the Kubernetes release, and CAPI also followed them.
What are the potential drawbacks of this? 
Keep creating betas until we have the code freeze and release RCs afterward.

Comment (tuminoid) [reply .github/workflows/release.yaml:48]: It is just commitment we need to agree to. 

We have not been doing too many branching RCs in the past, and in general not many alpha/betas either. We then can't do RC releases from main, but must cut the branch and fork the work at that point. We know branching for minor is not trivial with out multi-repo setup, so I just want us to make informed decision that this piece of code will dictate that process going forward.

For example, pushing a branch here does not mean we can test it in anyway unlike CAPI in a single repository setup.

File: .github/workflows/release.yaml
Comment (tuminoid) [reply .github/workflows/release.yaml]: seems fixed now

File: .github/workflows/release.yaml
Comment (tuminoid) [reply .github/workflows/release.yaml]: As discussed offline, let's mark this
/hold
For agreement next week when people are back from kubecon etc

**Review by Sunnatillo — APPROVED** [review]
/approve

File: .github/workflows/release.yaml
Comment (mquhuy) [inline .github/workflows/release.yaml:112]: Don't downgrade this

File: .github/workflows/release.yaml
Comment (mquhuy) [inline .github/workflows/release.yaml:24]: Don't downgrade this?

File: .github/workflows/release.yaml
Comment (mquhuy) [inline .github/workflows/release.yaml:19]: Maybe keep this (and add it to the other job too)? We may never push to forks' main, but who knows :-?

File: .github/workflows/release.yaml
Comment (mquhuy) [inline .github/workflows/release.yaml:27]: This action is also downgraded.

**Review by mquhuy — COMMENTED** [review]
Thanks @adilGhaffarDev , this looks like a very good change!

I had some nits from initial reviews, may come back with more later if I have time.

**Comment by kashifest** [comment]
/hold cancel
ready for review

File: docs/releasing.md
Comment (tuminoid) [inline docs/releasing.md:1]: This needs more revamping. Permissions section for example is now invalid. Reposotiry setup also needs some cleaning.

Release notes section also needs complete rewrite.

File: docs/releasing.md
Comment (tuminoid) [inline docs/releasing.md:78]: This is also not right.

File: docs/releasing.md
Comment (tuminoid) [inline docs/releasing.md:97]: This is also obsolete now, or needs to be reworded in a way that explains how the new automation does it.

**Review by tuminoid — COMMENTED** [review]
/hold

We need to have documents in more finished state. I realize we need further tuning after we've done a release, but at least the steps should be corrected. Now there is too much old stuff that should not be done by release team at all.

File: docs/releasing.md
Comment (adilGhaffarDev) [reply docs/releasing.md:1]: agree, I will make these changes.

File: docs/releasing.md
Comment (tuminoid) [inline docs/releasing.md:74]: There is no action with this name. All were adding is `release.yml` with `Create release` name.

What you're referring are jobs in the workflow.

File: docs/releasing.md
Comment (tuminoid) [inline docs/releasing.md:82]: Move this above build-images-action so jobs within same workflow are discussed together.

File: docs/releasing.md
Comment (adilGhaffarDev) [reply docs/releasing.md]: fixed.

**Comment by tuminoid** [comment]
/override metal3-centos-e2e-integration-test-main metal3-ubuntu-e2e-integration-test-main

**Review by tuminoid — COMMENTED** [review]
/lgtm

Lets take this baby for a test ride!

**Pull request merged**
